widget/entry.go

```diff
@@ -818,15 +837,9 @@ func (e *Entry) textProvider() *textProvider {
 		return e.text
 	}
 
-	e.Lock()
-	defer e.Unlock()
-	if e.text != nil {
-		return e.text
-	}
-
 	text := newTextProvider(e.Text, e)
-	text.ExtendBaseWidget(&text)
-	e.text = &text
+	text.ExtendBaseWidget(text)
+	e.text = text
```

Comment:
Is this safe? Shouldn't this and placeholderProvider() be set with SetFields?

Reply:
This is called so frequently and also initiates other code if a new one is created that making it lock itself was too deadlocky - so I opted for it must be locked outside this method, which as you noted above I have missed a few. They are fixed now so this doesn't have to be.

Reply:
`must be locked outside this method`

Read or Write lock?